### PR TITLE
Issue #720: enable to define num_parallel_calls in config

### LINKED ISSUE #720: Add option to limit number of CPUs for data preprocessing
## Subject of the feature

Currently, we are using AUTOTUNE in data preprocessing, e.g. https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/dataset/loader/interface.py#L113

However, it may take too many CPUs and thus also memories and this is not ideal on clusters.
Therefore we need to be able to configure this `num_parallel_calls`.

The fix can be,

- when loading config, we will set `num_parallel_calls` to the given value if provided, otherwise `tf.data.experimental.AUTOTUNE`.
- then we pass this `num_parallel_calls` to all funcs using it.

FYI @YipengHu @zacbaum @fepegar

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Fixes #720

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Documentation Update (fix or improvement on the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by YipengHu — APPROVED** [review]

**Comment by fepegar** [comment]
I know very little about parallel processing, but I have some comments.

1. Is `num_cpus` descriptive? I would say CPU cores, so maybe `num_cores`. PyTorch uses `num_workers`. I'm not sure it's equivalent, though.
2. Is it a good idea to use all of them by default? What about disabling multiprocessing by default and allowing the user to enable it by choosing >1 cores? Again, as a reference, in PyTorch, if `num_workers` is 0, multiprocessing will be disabled. Users have to explicitly define `num_workers` > 0. I haven't seen memory issues in the DGX (since February), but I think when 1 (or more!) users are training simultaneously using all cores, everything (even `ls`) is much slower. If users set e.g. 10 cores/GPU, I suspect things would run faster for everyone, and this is more likely to happen if users have to manually set up the multiprocessing settings.

**Comment by zacbaum** [comment]
> I know very little about parallel processing, but I have some comments.
> 
> 1. Is `num_cpus` descriptive? I would say CPU cores, so maybe `num_cores`. PyTorch uses `num_workers`. I'm not sure it's equivalent, though.

In TF (at least for what we're doing here) it is the number of scheduled cores that are used - so `num_cores` is likely better.

> 2. Is it a good idea to use all of them by default? What about disabling multiprocessing by default and allowing the user to enable it by choosing >1 cores? Again, as a reference, in PyTorch, if `num_workers` is 0, multiprocessing will be disabled. Users have to explicitly define `num_workers` > 0. I haven't seen memory issues in the DGX (since February), but I think when 1 (or more!) users are training simultaneously using all cores, everything (even `ls`) is much slower. If users set e.g. 10 cores/GPU, I suspect things would run faster for everyone, and this is more likely to happen if users have to manually set up the multiprocessing settings.

It's definitely not a good standard - especially where most of us have been using deepreg (using shared systems), as it can lock things up if the machine has low cores / low RAM in general (as is the case with the PTs). In DeepReg, setting the available cores for the data loaders will not change the amount of multiprocessing, etc. that occurs by changing `num_workers` in `model.fit()`. These can be set independently.

*Aside*: The reason `ls` will be slower is due to RAM being over-allocated. Metadata for these commands is cached in RAM, so when that runs out, the OS has to pull things from disk.

**Comment by fepegar** [comment]
Thanks, @zacbaum.

> Aside: The reason ls will be slower is due to RAM being over-allocated. Metadata for these commands is cached in RAM, so when that runs out, the OS has to pull things from disk.

Are you sure? I've noticed this when all cores are being 100% used, but there's a lot of RAM still available.

**Comment by zacbaum** [comment]
> Are you sure? I've noticed this when all cores are being 100% used, but there's a lot of RAM still available.

Well if there's no RAM it will be, but it could also be compounded by lack of CPU - Entirely possible! And also likely if CPU util > 90% or so.

**Comment by mathpluscode** [comment]
@zacbaum @fepegar Hi, I've renamed the vars and changed the default behaviour :) please approve if it looks ok, thx!

File: config/test/affine.yaml
Comment (zacbaum) [inline config/test/affine.yaml:21]: "workers" is a bit vague... Can we clarify what this actually does for the layperson?

File: deepreg/dataset/loader/interface.py
Comment (zacbaum) [inline deepreg/dataset/loader/interface.py:104]: especially that here we are now not consistent with the wording in the above :)

**Review by zacbaum — DISMISSED** [review]
We need to be consistent throughout - things refer to parallel calls, CPUs, workers; to someone who has no idea what this means, this isn't helpful.

Can we add some more detail to the docs for this to also say why someone may use this? (e.g. shared resource and don't want to leverage all the resource ;))

File: config/test/affine.yaml
Comment (mathpluscode) [reply config/test/affine.yaml:21]: I wish I know exactly what it is.

Here's tf doc
- https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map
- https://www.tensorflow.org/guide/data_performance#parallelizing_data_transformation

I will try to clarify it a bit further. If still not ok, feel free to propose one or modify it on the branch directly.

**Comment by mathpluscode** [comment]
> We need to be consistent throughout - things refer to parallel calls, CPUs, workers; to someone who has no idea what this means, this isn't helpful.
> 
> Can we add some more detail to the docs for this to also say why someone may use this? (e.g. shared resource and don't want to leverage all the resource ;))

I tried again :)

```
number elements to process asynchronously in parallel
during preprocessing, -1 means unlimited, heuristically it should be set to
the number of CPU cores available. AUTOTUNE=-1 means not limited.
```

**Comment by mathpluscode** [comment]
@zacbaum :0 I'm still waiting for approvals :)

**Comment by fepegar** [comment]
@mathpluscode By default, all CPU cores are used?

File: config/test/affine.yaml
Comment (zacbaum) [inline config/test/affine.yaml:21]: Why should this be the number of available cores?

File: deepreg/dataset/loader/interface.py
Comment (zacbaum) [inline deepreg/dataset/loader/interface.py:92]: As @fepegar has mentioned, are we sure we want this to be the default behaviour?

File: deepreg/train.py
Comment (zacbaum) [inline deepreg/train.py:65]: I would expect that num_workers is also used for model.fit if we are using the same name here... Which doesn't appear to be the case.

File: docs/source/docs/cli.md
Comment (zacbaum) [inline docs/source/docs/cli.md:43]: What happens if users enter 0, or a number higher than the CPU cores they have? Do we need to check for this if this behaviour causes issues?

**Comment by zacbaum** [comment]
> @zacbaum :0 I'm still waiting for approvals :)

@mathpluscode, I have left a few comments --

I (along with many others) was not working during the Easter break, and certainly not at 11:56pm on a Friday when you tagged me, so you'll have to forgive the wait...

File: config/test/affine.yaml
Comment (mathpluscode) [reply config/test/affine.yaml:21]: TensorFlow said so. will update in the docstring

File: deepreg/train.py
Comment (mathpluscode) [reply deepreg/train.py:65]: image
We are using tf.data therefore this argument is not used

File: docs/source/docs/cli.md
Comment (mathpluscode) [reply docs/source/docs/cli.md:43]: I only limit the CPU if this number if > 0, thus I will add an error

File: deepreg/dataset/loader/interface.py
Comment (mathpluscode) [reply deepreg/dataset/loader/interface.py:92]: ok then i change to 1

Comment (mathpluscode) [reply deepreg/dataset/loader/interface.py:95]: actually, I believe `num_workers` limits the CPU globally, and this will not leads to use more CPUs, I just tested.

**Comment by mathpluscode** [comment]
> @mathpluscode By default, all CPU cores are used?

By default, CPU is limited to 1 at the very beginning with the definition of the environment variables. So the num_parallel_call will be limited by that.

I insist on not setting both to 1, as if the user does not want any limit, he/she has to modify two values. Now, you can limit the CPUs with one single value, as another one is bounded by the previous one.

**Review by fepegar — APPROVED** [review]

**Pull request merged**
